### PR TITLE
[Enhancement] support memory scratch sink in pipeline

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    return _is_finished;
+}
+
+Status MemoryScratchSinkOperator::set_finishing(RuntimeState* state) {
+    _is_finished = true;
+    return Status::OK();
+}
+
+bool MemoryScratchSinkOperator::pending_finish() const {
+    const_cast<MemoryScratchSinkOperator*>(this)->try_to_put_sentinel();
```
Comment (satanson) [inline be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp:40]: add some comment to explain why MemoryScratchSinkOperator needs pending_finish.

File: be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    return !(_is_finished && _pending_result == nullptr && _has_put_sentinel);
+}
+
+Status MemoryScratchSinkOperator::set_cancelled(RuntimeState* state) {
```
Comment (satanson) [inline be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp:44]: add some comment to explain why MemoryScratchSinkOperator needs to override set_cancelled, at the present time, only ExchangeSink overrides this function to handle brpc errors, other operators' set_cancelled is trivial and it is equivallent to set_finished function.

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 3]
```diff
@@ -805,6 +807,15 @@ Status FragmentExecutor::_decompose_data_sink_to_operator(RuntimeState* runtime_
                 context->next_operator_id(), request.output_sink().mysql_table_sink,
                 mysql_table_sink->get_output_expr(), dop, fragment_ctx);
         fragment_ctx->pipelines().back()->add_op_factory(op);
+    } else if (typeid(*datasink) == typeid(starrocks::MemoryScratchSink)) {
+        auto* memory_scratch_sink = down_cast<starrocks::MemoryScratchSink*>(datasink.get());
```
Comment (satanson) [inline be/src/exec/pipeline/fragment_executor.cpp:811]: Can memory_scratch_sink be parallelized ? in non-pipeline engine, is this Sink one-instance?  maybe multiple memory_scratch_sinks can emit  RecordBatches on the same wire concurrently.

Comment (silverbullet233) [reply be/src/exec/pipeline/fragment_executor.cpp:811]: > Can memory_scratch_sink be parallelized ? in non-pipeline engine, is this Sink one-instance? maybe multiple memory_scratch_sinks can emit RecordBatches on the same wire concurrently.

This sink can be parallelized, but it is determined by flink/spark. for example, if SR has multiple partitions and flink has multile source tasks, flink will decide which task is responsible for reading which data. The each source task will open a scanner to obtain data. for BE, each scanner is equivalent to a fragment to be executed. so we don't need to consider the parallelism problem.

File: be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp:42]: done

File: be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/sink/memory_scratch_sink_operator.cpp:46]: done

**Review by satanson — DISMISSED** [review]

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by kangkaisen — DISMISSED** [review]

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by satanson — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**